apps/emqx_gateway/test/emqx_gateway_authn_http_SUITE.erl

```diff
@@ -0,0 +1,379 @@
+%% limitations under the License.
+%%--------------------------------------------------------------------
+
+-module(emqx_gateway_authn_http_SUITE).
+
+-compile(nowarn_export_all).
+-compile(export_all).
+
+-include("emqx_authn.hrl").
+-include_lib("eunit/include/eunit.hrl").
+-include_lib("common_test/include/ct.hrl").
+-include_lib("emqx/include/emqx_placeholder.hrl").
+
+-define(PATH, [?CONF_NS_ATOM]).
+
+-define(HTTP_PORT, 33333).
+-define(HTTP_PATH, "/auth").
+
+-define(checkMatch(Guard),
+    (fun(Expr) ->
+        case (Expr) of
+            Guard ->
+                ok;
+            X__V ->
+                erlang:error(
+                    {assertMatch, [
+                        {module, ?MODULE},
+                        {line, ?LINE},
+                        {expression, (??Expr)},
+                        {pattern, (??Guard)},
+                        {value, X__V}
+                    ]}
+                )
+        end
+    end)
+).
+-define(FUNCTOR(Expr), fun() -> Expr end).
+-define(FUNCTOR(Arg, Expr), fun(Arg) -> Expr end).
+
+-define(PROTOCOLS, [coap, lwm2m, 'mqtt-sn', stomp, exproto]).
+
+-define(CONFS, [
+    emqx_coap_SUITE,
+    emqx_lwm2m_SUITE,
+    emqx_sn_protocol_SUITE,
+    emqx_stomp_SUITE,
+    emqx_exproto_SUITE
+]).
+
+-define(CASES, [
+    fun case_coap/0,
+    fun case_lwm2m/0,
+    fun case_emqx_sn/0,
+    fun case_stomp/0,
+    fun case_exproto/0
+]).
+
+-define(AUTHNS, [fun set_http_authn/1]).
+
+-type auth_controller() :: fun((start | stop) -> ok).
+
+all() ->
+    emqx_common_test_helpers:all(?MODULE).
+
+init_per_suite(Config) ->
+    _ = application:load(emqx_conf),
+    ok = emqx_common_test_helpers:load_config(emqx_gateway_schema, init_conf()),
+    emqx_common_test_helpers:start_apps([emqx_authn, emqx_gateway]),
+    application:ensure_all_started(cowboy),
+    Config.
+
+end_per_suite(_) ->
+    clear_authn(),
+    ok = emqx_common_test_helpers:load_config(emqx_gateway_schema, <<>>),
+    emqx_common_test_helpers:stop_apps([emqx_authn, emqx_gateway]),
+    application:stop(cowboy),
+    ok.
+
+init_per_testcase(_Case, Config) ->
+    {ok, _} = emqx_cluster_rpc:start_link(node(), emqx_cluster_rpc, 1000),
+    Config.
+
+end_per_testcase(_Case, Config) ->
+    Config.
+
+%%------------------------------------------------------------------------------
+%% Tests
+%%------------------------------------------------------------------------------
+t_authn(_) ->
+    test_gateway_with_auths(?CASES, ?AUTHNS).
+
+%%------------------------------------------------------------------------------
+%% Tests
+%%------------------------------------------------------------------------------
+
+case_coap() ->
+    Login = fun(URI, Checker) ->
+        Action = fun(Channel) ->
+            Req = emqx_coap_SUITE:make_req(post),
+            Checker(emqx_coap_SUITE:do_request(Channel, URI, Req))
+        end,
+        emqx_coap_SUITE:do(Action)
+    end,
+    Prefix = emqx_coap_SUITE:mqtt_prefix(),
```

Comment:
These calls seem to be poorly maintained.
Can we encapsulate all these client-side-related operations in a separate module? i.e: `emqx_gateway_ct_coap_client.erl`

Reply:
I think we can't make everything better in one PR, the optimisation of clients maybe be another task